src/camlzip/iostream_camlzip.ml

```diff
@@ -21,93 +21,141 @@ type mode =
   | Inflate
   | Deflate of int
 
-class transduce_in_ ~mode (ic : #In_buf.t) : In.t =
-  let zlib_str =
-    match mode with
-    | Inflate -> Zlib.inflate_init false
-    | Deflate lvl -> Zlib.deflate_init lvl false
-  in
-  let state = ref In_progress in
-  object
-    method close () =
-      (match mode with
-      | Inflate -> Zlib.inflate_end zlib_str
-      | Deflate _ -> Zlib.deflate_end zlib_str);
-      In.close ic
-
-    method input buf i len =
-      let n_written = ref 0 in
-
-      while !n_written = 0 && !state != Done do
-        match !state with
-        | Done -> assert false
-        | In_progress ->
-          let islice = In_buf.fill_buf ic in
-          if islice.len = 0 then
-            state := Consuming_rest
-          else (
-            let finished, used_in, used_out =
-              (match mode with
-              | Inflate -> Zlib.inflate
-              | Deflate _ -> Zlib.deflate)
-                zlib_str islice.bytes islice.off islice.len buf i len
-                Zlib.Z_NO_FLUSH
-            in
-            if finished then state := Done;
-            In_buf.consume ic used_in;
-            n_written := used_out
-          )
-        | Consuming_rest ->
-          (* finish sending the internal state *)
-          let islice = Slice.empty in
+module In_trans = struct
+  type t =
+    | St : {
+        ic: 'ic;
+        mode: mode;
+        mutable state: transduce_state;
+        zlib_str: Zlib.stream;
+        close: 'ic -> unit;
+        fill_buf: deadline:float option -> 'ic -> Slice.t;
+      }
+        -> t
+
+  let of_in ~mode (ic : #In_buf.t) : t =
+    let zlib_str =
+      match mode with
+      | Inflate -> Zlib.inflate_init false
+      | Deflate lvl -> Zlib.deflate_init lvl false
+    in
+    St
+      {
+        state = In_progress;
+        mode;
+        zlib_str;
+        ic;
+        close = In_buf.close;
+        fill_buf =
+          (fun ~deadline ic ->
+            assert (deadline = None);
```

Comment:
`ic` itself doesn't have a deadline, this particular assertion is mostly there because we read from a `In_buf.t` which doesn't support timeouts at all, but the common state allows for a deadline. There will never be a deadline on this code path, hence the assertion.